### DISCUSSION TITLE
ussl module for Raspberry Pico W

### DISCUSSION THREAD

**Discussion by IanW6374** [micropython/micropython#9222] [Libraries & Drivers]
Hi 

I am trying to create a TLS connection to a remote server (Pico W with MicroPython) but having issue trying to define the ssl parameters. The ssl parameters are stored in the "sslparams" dictionary object but I get "TypeError: extra keyword arguments given" if I try anything other than None for the certificates.

I have tried loading the certificate into a variable and defining the certificate path to no avail - Is there any information on what is supported on the RP2 port of ussl??

Any tips would be gratefully received :)

Sample code is attached

[samplecode.txt](https://github.com/micropython/micropython/files/9495868/samplecode.txt)

**Comment by jimmo** [comment]
@IanW6374 the cert validation is still under development and is not supported on the release version (1.19.1). There is some support in the nightly/unstable versions.

Which version of MicroPython are you using?

Which documentation are you following? See https://docs.micropython.org/en/latest/library/ssl.html for valid arguments to wrap_socket (note that this doesn't quite match CPython, we're working on adding support for a CPython-compatible SSLContext).

See also example here:
https://github.com/micropython/micropython/blob/master/tests/net_inet/ssl_cert.py

**Reply by IanW6374** [reply]
@jimmo Thanks for your prompt response

I am using MicroPython v1.19.1 on 2022-09-06; Raspberry Pi Pico W with RP2040 (rp2-pico-w-20220906-unstable-v1.19.1-375-ge90b85cc9.uf2)

I was using the MIcroPython ssl module documentation and using the following  sslparams = {'server_side':False, 'keyfile':None, 'certfile':None, 'ca_certs':None, 'cert_reqs':ussl.CERT_NONE} because If I use the "cadata" keyword I get a "TypeError: unexpected keyword argument 'cadata'" so I swapped for "ca_certs:None" which seemed to be accepted.

Are all the certs / key going to be required in DER format?

**Reply by Carglglz** [reply]
> I was using the MIcroPython ssl module documentation and using the following sslparams = {'server_side':False, 'keyfile':None, 'certfile':None, 'ca_certs':None, 'cert_reqs':ussl.CERT_NONE} because If I use the "cadata" keyword I get a "TypeError: unexpected keyword argument 'cadata'" so I swapped for "ca_certs:None" which seemed to be accepted.

@IanW6374 maybe you are using `ssl` module from `micropython-lib/ssl` which has not been yet updated? Try to use 
``` python
import ussl as ssl 
```
And use `cadata` keyword, that should work. 
 
> Are all the certs / key going to be required in DER format?

For now yes, you can convert them using `openssl` , e.g.
```console
$ openssl x509 -in cert.pem -out cert.der -outform DER
```

From your `samplecode.txt`

```
sslparams = {'server_side':False, 'keyfile':key, 'certfile':None, 'cert_reqs':ssl.CERT_NONE, 'ca_certs':None}
```
Client side does not require key/cert unless the server wants to perform client authentication. (in which case you need both key and cert).
`cadata` is to validate the server you are connecting to, which should be the default setting to use, .i.e `cert_reqs=ssl.CERT_REQUIRED`

Based in your code you would want to do:
Mutual authentication
```
sslparams = {'server_side':False, 'keyfile':key, 'certfile':cert, 'cert_reqs':ssl.CERT_REQUIRED, 'cadata':ca}

```
or Server authentication
```
sslparams = {'server_side':False, 'keyfile':None, 'certfile':None, 'cert_reqs':ssl.CERT_REQUIRED, 'cadata':ca}

```

Also note that in Raspberry Pico W certificate validation requires the Pico to have the correct RTC time. (check #9071)

**Reply by IanW6374** [reply]
@Carglglz Thanks for the information - I think I was importing the ssl not ussl module 

I have the example "https://github.com/micropython/micropython/blob/master/tests/net_inet/ssl_cert.py" now working correctly with 

sslparams = {'server_side':False, 'cert_reqs':ssl.CERT_REQUIRED, 'cadata':ca_cert_chain}

however if I include keyfile and certfile keywords i.e.

sslparams = {'server_side':False, 'keyfile':None, 'certfile':None, 'cert_reqs':ssl.CERT_REQUIRED, 'cadata':ca_cert_chain}

then I get the error "TypeError: extra keyword arguments given"

**Reply by Carglglz** [reply]
> sslparams = {'server_side':False, 'keyfile':None, 'certfile':None, 'cert_reqs':ssl.CERT_REQUIRED, 'cadata':ca_cert_chain}
then I get the error "TypeError: extra keyword arguments given"

@IanW6374 yes, sorry about that, right now it is a bit confusing, `ussl` uses `key` and `cert` keywords instead of `keyfile` and `certfile`, for  the next release this should be fixed I guess, so it accepts `keyfile`, `certfile` too. 👍🏼 (Which is the version I'm working on, that's why I used those keywords 😅)

So for now use 
```
sslparams = {'server_side':False, 'key':None, 'cert':None, 'cert_reqs':ssl.CERT_REQUIRED, 'cadata':ca_cert_chain}
```

**Reply by IanW6374** [reply]
@Carglglz All working now - Thanks for the prompt assistance :)

**Reply by jimmo** [reply]
Thanks @Carglglz 

We'll try and sort out the SSLContext PR soon (there's a lot on our plate right now). Also need to figure out what to do w.r.t. the one in micropython-lib.

**Reply by JMFUS** [reply]
Hello
I'm trying to connect to my Filezilla Server with FTP_TLS
My Micropython version is MicroPython v1.22.0-preview.263.g0e706a62b on 2023-12-18; Raspberry Pi Pico W with RP2040
I'm using ussl and ftplibtls.py (Christopher Arndt) librairies
I can connect to my server but when I try to display the files list, I have the following error :

```
========================== CONNEXION AU FTP ==========================
line =  b'220-FileZilla Server 0.9.60 beta\r\n'
line =  b'220-written by Tim Kosse (tim.kosse@filezilla-project.org)\r\n'
line =  b'220 Please visit https://filezilla-project.org/\r\n'
line =  b'234 Using authentication type TLS\r\n'
line =  b'331 Password required for jeanmarcf\r\n'
line =  b'230 Logged on\r\n'
line =  b'200 PBSZ=0\r\n'
line =  b'200 Protection level set to P\r\n'
line =  b'200 Type set to A\r\n'
line =  b'227 Entering Passive Mode (88,122,12,99,128,233)\r\n'
line =  b'150 Opening data channel for directory listing of "/"\r\n'
resp : 150 Opening data channel for directory listing of "/"
Traceback (most recent call last):
  File "<stdin>", line 128, in <module>
  File "ftplib.py", line 509, in retrlines
  File "ftplibtls.py", line 140, in ntransfercmd
OSError: stream operation not supported
```

Here is the function where the error appears :

```py
    def ntransfercmd(self, cmd, rest=None):
        conn, size = super().ntransfercmd(cmd, rest)
        if self._prot_p:
            sock = getattr(conn, '_sock', conn)
#             sock = ssl.wrap_socket(conn,
#                                    keyfile=self.keyfile,
#                                    certfile=self.certfile,
#                                    cert_reqs=self.cert_reqs)
            sock = ssl.wrap_socket(conn,
                                   key=self.keyfile,
                                   cert=self.certfile,
                                   cert_reqs=self.cert_reqs)
 
            if hasattr(conn, '_sock'):
                conn._sock = sock
            else:
                conn = sock

        return conn, size
```

Somebody can help me ?
Thx in advance

(Edit by @jimmo to fix formatting)

**Reply by jimmo** [reply]
@JMFUS please see [Welcome (Please read before posting)](https://github.com/orgs/micropython/discussions/9111) for how to format code in forum posts.

**Reply by JMFUS** [reply]
Sorry
Thx for the modifications

**Reply by JMFUS** [reply]
Hello 
I'm still working on my FTP over TLS connection for my Raspberry Pi Pico WH
I have a Filezilla server 1.8.0 with a certificate
I copied this certificate on my Raspberry
And the code i'm using is :
```    
    print('Loading CA Certificate')
    with open(cert_file_path, 'rb') as f:
        cacert = f.read()
    f.close()
    
    ftp_conn = FTP_TLS(keyfile=None,certfile=cert_file_path,cert_reqs=ssl.CERT_REQUIRED,ca_certs=cacert,server_hostname=adresse_serveur)
    
    print("Connexion au serveur FTP...")
    try:
        ftp_conn.connect(adresse_serveur,port)
    except Exception as e:
        print("Erreur lors de la connexion au serveur FTP :", str(e))
    
    try:    
        print("Connexion établie. Tentative d'authentification...")
        ftp_conn.login(user=utilisateur_OST, passwd=mot_de_passe_OST)
    except Exception as e:
        raise e
    
    try: 
        ftp_conn.prot_p()
    except Exception as e:
        raise e

    ftp_conn.cwd("/")

    print("============================ CONTENU FTP ============================")
    items = FTP_file_exist(ftp_conn)
    print("items = ",items)
    print("========================= FIN CONNEXION FTP =========================")
    ftp_conn.quit()
```
    
But now, the reply is "invalid cert"
  Traceback (most recent call last):
  File "<stdin>", line 168, in <module>
  File "<stdin>", line 166, in <module>
  File "ftplibtls.py", line 78, in login
  File "ftplibtls.py", line 101, in auth
  File "ftplibtls.py", line 99, in auth
  ValueError: invalid cert

Is there somebody who can help me ?

Thx in advance
And happy new year for all

**Reply by sophiedegran** [reply]
Hi
Did you convert key/cert to DER format ? (with openssl)

**Reply by JMFUS** [reply]
Sorry
I'm a newby :)
The only file generated by FZ is trustedcerts.xml located on the server in users/toto/appdata/roaming/filezilla/trustedcerts.xml :

```
<FileZilla3 version="3.66.4" platform="windows">
	<TrustedCerts>
		<Certificate>		<Data>3082019e30820144a003020102021429ae7a3c957c889894061fbbd4ebcf5afc2e1b90300a06082a8648ce3d04030230333131302f0603550403132866696c657a696c6c612d7365727665722073656c66207369676e6564206365727469666963617465301e170d3233313232373130353330395a170d3234313232373130353830395a30333131302f0603550403132866696c657a696c6c612d7365727665722073656c66207369676e65642063657274696669636174653059301306072a8648ce3d020106082a8648ce3d03010703420004300daf8be3ce8a54c212c1dc6fcbd5fa99938011bc526e7543064b1fba787d2e56640b9a54981feaefe83883e1a21fbc90161a2305332c969933c459311d9a0ea336303430140603551d11040d300b82096c6f63616c686f7374300e0603551d0f0101ff0404030205a0300c0603551d130101ff04023000300a06082a8648ce3d040302034800304502207629246b6c2261426a38d42bccde4547d1c407683cbf85123ed0877274e97385022100fd727434efbef158cad6b2bf3da8746d0a6dfb8725ad627f6dc3bad4547c96e5</Data>
			<ActivationTime>1703674389</ActivationTime>
			<ExpirationTime>1735297089</ExpirationTime>
			<Host>127.0.0.1</Host>
			<Port>22</Port>
			<TrustSANs>0</TrustSANs>
		</Certificate>
	</TrustedCerts>
	<FtpSessionResumption>
		<Entry Host="192.168.0.197" Port="21">true</Entry>
		<Entry Host="127.0.0.1" Port="22">true</Entry>
	</FtpSessionResumption>
</FileZilla3>
```
I don' know what to do with that
I just copied it on my Raspberry

**Reply by sophiedegran** [reply]
Ok don't know about filezilla but have a look on the wiki or the forum of filezilla you will have the answer...
After don't forget to convert to DER and to have the ip adress define in your cert (192.168.0.x)  and the date/time of your client near the server..
I'm sure you will do it ;)

**Comment by JMFUS** [comment]
ThxI'll checkThxEnvoyé depuis mon appareil Galaxy
-------- Message d'origine --------De : sophiedegran ***@***.***> Date : 30/12/2023  21:19  (GMT+01:00) À : micropython/micropython ***@***.***> Cc : JMFUS ***@***.***>, Mention ***@***.***> Objet : Re: [micropython/micropython] ussl module for Raspberry Pico W (Discussion #9222) 
Ok don't know about filezilla but have a look on the wiki or the forum of filezilla you will have the answer...
After don't forger to convert to DER and to have the ip adress define in your cert
I'm sure you will do it ;)

**Comment by JMFUS** [comment]
Hello
Now I fixed the certificate problem and I can connect to my server with TLS 
And I find another closed door :
When I send the command LIST to the server to have the list of the files , I have now the error :

resp : 150 Starting data transfer.
Traceback (most recent call last):
  File "<stdin>", line 202, in <module>
  File "ftplib.py", line 508, in retrlines
  File "ftplibtls.py", line 159, in ntransfercmd
OSError: stream operation not supported

The problem is in this function :

```
  def ntransfercmd(self, cmd, rest=None):
        conn, size = super().ntransfercmd(cmd, rest)
        if self._prot_p:
            sock = getattr(conn, '_sock', conn)

            sock = ussl.wrap_socket(conn,
                                   key=self.keyfile,
                                   cert=self.certfile,
                                   cert_reqs=self.cert_reqs)
            
            if hasattr(conn, '_sock'):
                conn._sock = sock
            else:
                conn = sock

        return conn, size
```
      
in the ftplibtls library I'm using, at the line :

sock = ussl.wrap_socket(conn, ........

An idea?

Thx
